### ISSUE TITLE
Containers disapearing

### ISSUE THREAD

**Issue by Maikel-J** [containrrr/watchtower#868]
**Describe the bug**
One of my containers (sonarr) keeps disapearing after watchtower updates it.

**Environment**
 - synology NAS (ds920+)
 - Docker version 18.09.0-0513

<details>
 <summary><b> Logs from running watchtower </b></summary>

```
time="2021-03-25T15:43:56Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-26 15:43:56 +0000 UTC m=+86400.382600350"
time="2021-03-25T17:26:30Z" level=info msg="Waiting for running update to be finished..."
time="2021-03-25T17:26:36Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-26 17:26:36 +0000 UTC m=+86400.573183055"
time="2021-03-26T17:27:31Z" level=info msg="Found new linuxserver/sonarr:latest image (sha256:0294b369af569d44c8ddd7277fda95776a2bbe94bca17705ff63d6b649100c4c)"
time="2021-03-26T17:28:02Z" level=info msg="Found new linuxserver/jackett:latest image (sha256:8a5fcb80d73ac11d79d194f9ce2a3b73f156374e25463dead96a762f06d0e444)"
time="2021-03-26T17:28:52Z" level=info msg="Found new plexinc/pms-docker:latest image (sha256:1dfb8b3763b4c697ab94cde0838bd7962a540594121af6f582331c2ab7580d65)"
time="2021-03-26T17:28:52Z" level=info msg="Stopping /jackett (2e2696d0acd7fffd53c949ea280eec07a69a455a6d8248c77192a3a32b41a3f5) with SIGTERM"
time="2021-03-26T17:29:02Z" level=info msg="Stopping /sonarr (219b3159d8fe0b7f75bb7488f2e2c1a22980c16024d079c4e71d69d0dc400d9a) with SIGTERM"
time="2021-03-26T17:29:11Z" level=info msg="Stopping /plexinc (3e9ce0f3dce855552c7049584eaa3a36245e15cdcb5eb3a789c329203329c453) with SIGTERM"
time="2021-03-26T17:29:24Z" level=info msg="Creating /plexinc"
time="2021-03-26T17:29:28Z" level=info msg="Creating /sonarr"
time="2021-03-26T17:29:30Z" level=error msg="Error response from daemon: could not get container for jackett: No such container: jackett"
time="2021-03-26T17:29:30Z" level=info msg="Creating /jackett"
time="2021-03-27T11:21:24Z" level=info msg="Waiting for running update to be finished..."
time="2021-03-27T11:33:12Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-28 11:33:12 +0000 UTC m=+86415.466201597"
time="2021-03-27T13:57:07Z" level=info msg="Waiting for running update to be finished..."
time="2021-03-27T14:00:18Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-28 14:00:18 +0000 UTC m=+86435.687601620"
time="2021-03-27T14:51:34Z" level=info msg="Waiting for running update to be finished..."
time="2021-03-27T14:51:55Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-28 14:51:55 +0000 UTC m=+86400.208564888"
time="2021-03-27T18:16:33Z" level=info msg="Waiting for running update to be finished..."
time="2021-03-27T18:19:47Z" level=info msg="Starting Watchtower and scheduling first run: 2021-03-28 18:19:47 +0000 UTC m=+86427.580408417"
time="2021-03-28T18:20:33Z" level=info msg="Found new linuxserver/sonarr:latest image (sha256:f1f4de291b902159a62053e2f08435f3aa19e81364006c1b70461494e1a8fed9)"
time="2021-03-28T18:20:51Z" level=info msg="Found new linuxserver/jackett:latest image (sha256:e3545772860553a0ab4d86b3395b8e20a85eda5ac08cae6ee1bbaf72889bdc21)"
time="2021-03-28T18:21:11Z" level=info msg="Found new haugene/transmission-openvpn:latest image (sha256:c877eff88de26f49ca7da4091748f6d14c835f903d5644fc61c0a83b385c3c9e)"
time="2021-03-28T18:21:11Z" level=info msg="Stopping /transmission-openvpn (a810cf0938228faf0a28cfa0de346b53760ff5e1bc4f4b3df1b37473a28ebdf9) with SIGTERM"
time="2021-03-28T18:21:19Z" level=info msg="Stopping /jackett (94bec911b8d713b956396dc64f9e09724eed8240aafca57c27456356f1386258) with SIGTERM"
time="2021-03-28T18:21:28Z" level=info msg="Stopping /sonarr (9f51e75b9bd3b4cfa0a14475845da3dc5395e4091c8cc8b010f55edf16f08b05) with SIGTERM"
time="2021-03-28T18:21:38Z" level=info msg="Creating /sonarr"
time="2021-03-28T18:21:41Z" level=error msg="Error response from daemon: could not get container for transmission-openvpn: No such container: transmission-openvpn"
time="2021-03-28T18:21:41Z" level=info msg="Creating /jackett"
time="2021-03-28T18:21:45Z" level=info msg="Creating /transmission-openvpn"
```

</details>

**Additional context**
sonarr is dependant on tranmission and jackett.

**Comment by simonaronsson** [comment]
How have you set up the dependencies? Cause it's not really stopping and starting it in any specific order, which would explain the issues you're experiencing.

**Comment by Maikel-J** [comment]
This is how it is setup for sonarr.
```
"links" : [
      {
         "alias" : "transmission",
         "link_container" : "transmission-openvpn"
      },
      {
         "alias" : "jackett",
         "link_container" : "jackett"
      }
   `]
```

**Comment by simonaronsson** [comment]
Not sure why it's not enforcing the dependency then. Would you mind having a try with https://containrrr.dev/watchtower/linked-containers/ to see whether that resolves your issue? Whichever the case I'd like to keep this issue open for further investigation.

**Comment by Maikel-J** [comment]
I am new to docker, so I am not sure how it all works but I this the right format for it?
```
"Labels": "com.centurylinklabs.watchtower.depends-on jackett, transmission_openvpn",
```
This is for the sonarr container.

**Comment by piksel** [comment]
The links should suffice (not that I am confident in how to properly set up links between containers). But the label should be
```
"Labels": "com.centurylinklabs.watchtower.depends-on=jackett,transmission_openvpn"
```
I **think**.

**Comment by juliansteenbakker** [comment]
I have the same problem here. Sonarr and Radarr keeps dissapearing with the following log message (log goes down to up): 
```
2021-04-06 08:38:51 | stderr | time="2021-04-06T08:38:51Z" level=info msg="Creating /Jackett"
2021-04-06 08:38:51 | stderr | time="2021-04-06T08:38:51Z" level=error msg="Error response from daemon: could not get container for Jackett: No such container: Jackett"
2021-04-06 08:38:50 | stderr | time="2021-04-06T08:38:50Z" level=info msg="Creating /Radarr"
2021-04-06 08:38:39 | stderr | time="2021-04-06T08:38:39Z" level=info msg="Stopping /Radarr (ecc44c2dc6ee62e1346af1ad145d9cf056ae0072a3182ad12ddb95f12823fa40) with SIGTERM"
2021-04-06 08:38:28 | stderr | time="2021-04-06T08:38:28Z" level=info msg="Stopping /Jackett (f9ab8075754881371dc5fa493911280c9e7a76bc86d9ce0db8032e19f2269d06) with SIGTERM"
2021-04-06 08:38:25 | stderr | time="2021-04-06T08:38:25Z" level=info msg="Found new linuxserver/radarr:latest image (sha256:a3c24f6bdd3debf145abe085aa28328346a80459ae4b529980bab27a63b4cafc)"
2021-04-06 08:38:04 | stderr | time="2021-04-06T08:38:04Z" level=info msg="Found new linuxserver/jackett:latest image (sha256:4082f80b6bf508ec97265a681e8397044313b14860203aba4f826a6a40ecf779)"
```

It seems like it's not detecting the linked containers and creating the linked one before the other has been made.

**Comment by simonaronsson** [comment]
We found an issue in the linking logic that seems to have been introduced a couple of months ago, which flipped the starting order around. This will likely have caused containers with links to disappear as watchtower would have attempted to create them prior to creating the container they depended on.

```go
	containersToUpdate := []container.Container{}
	if !params.MonitorOnly {
		for i := len(containers) - 1; i >= 0; i-- {
			if !containers[i].IsMonitorOnly() {
				containersToUpdate = append(containersToUpdate, containers[i])
			}
		}
	}
```
https://github.com/containrrr/watchtower/blame/4142f7966a6703ed3221f1ebffb6e03fa4e18747/internal/actions/update.go#L60-L67

A fix will be included in the next release, which we hope to release to you in the next couple of days.

**Comment by juliansteenbakker** [comment]
Awesome, thanks for the fast reply!

**Comment by simonaronsson** [comment]
In the meantime, if you want to try it out @juliansteenbakker, we're always happy to have people help with testing on the `latest-dev` tag. Just don't forget to move back once it's been released 😅

**Comment by juliansteenbakker** [comment]
Will do. I'll let it run for a few days and report back as soon as Sonarr or Radarr has gotten any update.

**Comment by juliansteenbakker** [comment]
After 12 days of running i can confirm updating is going well and the containers are not disapearing. However, there is still a problem with linking. It seems like it still references the old Jackett container. When i try to start Sonarr or Radarr it gives the following error even when Jackett is running:
![Schermafbeelding 2021-05-03 om 08 50 44](https://user-images.githubusercontent.com/20091319/116848338-b0b1b580-abec-11eb-9e69-a608cf91eb41.png)

When i undo and redo the links for Sonarr and Radarr everything works fine.

**Comment by Maikel-J** [comment]
Have there been any updates on this problem?

**Comment by Sveeeeeen** [comment]
Issue still exists.

I am using docker compose and "depends on".

**Comment by simonaronsson** [comment]
This should be solved in the latest version (see pr #1103) . If that is not the case, please open a new issue with logs as well as a minimal repro. Thanks!
